021-special-repository-for-st-clients-based-on-example-clients.md

```diff
@@ -0,0 +1,61 @@
+# Special repository for ST clients based on example clients
+
+This proposal suggests creating a new repository for `Strimzi` ST client based on 
+[Strimzi client-examples](https://github.com/strimzi/client-examples).
+
+## Current situation
+
+Currently, we are using two clients in our STs:
+ - `InternalKafkaClients` (based on `test-client` image) 
+ - `example clients` (from [Strimzi client-examples](https://github.com/strimzi/client-examples)).
+
+The plan is to remove the `InternalKafkaClients` and keep only `example clients`.
+
+## Motivation
+
+While testing `Strimzi` we need, in some cases, special configuration of clients, which is not implemented in the
+`client-examples`.
+How we discussed earlier, the `client-examples` repository should be really _exemplary_, and we should not add any extra
+_configuration_ or _extensions_ to it. For this kind of enhancements we should have repository, which will have
+`client-examples` as base, and we will be able to add special setting without disrupting the basic idea of example
+clients.
+
+## Proposal
+
+ * Create a new repository for `systemtest client`
+    * complex implementation of clients for testing
+    * we'll use both Kafka and Bridge clients from `client-examples`
+    * will be based on [Strimzi client-examples](https://github.com/strimzi/client-examples) - we'll copy the
+      `client-examples` code and then modify it - each repo will then _go their own way_
+    * we'll be able to modify it with our special configuration
+    * the main idea of example clients remain intact
+    
+ * The original `client-examples` repository will be kept
+
+## Advantages
+
+There are many things we can implement. 
+Good example is returning exceptions and return codes into the `job` status 
+(as we are using `k8s` jobs for deploying the example clients) and asserting it in tests - we have to grep exceptions
+from the job log at the moment - which can be a problem. 
+
+## Images and releases
+
+The images will be built as in `client-examples` after each merged PR and pushed to `strimzi` repository on `quay.io`.
+I would keep the `latest` tag, no need for extra releases and versioning.
+
+## Kafka version
+
+ST client will support and use the latest Kafka version supported by `Strimzi`.
```

Comment:
So if we will do Strimzi 0.22.2 and master will use Kafka 2.8, `systemtest-client` as well, how you will use `latest version supported by Strimzi` without additional versioning for `systemtest-client` ?

Reply:
I read this to mean it will always support a Kafka version that's at least as recent as that supported by the operator. There's no reason why it can't be more recent. And, there's actually some value in being able to test old clients against the new broker (which you could do by having multiple jars in the image and selecting them according to an env var).